app/src/main/java/com/example/healthinspector/LocationService.java

```diff
@@ -0,0 +1,127 @@
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.OnTokenCanceledListener;
+import com.google.android.gms.tasks.Task;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+public class LocationService extends Service {
+    LocationManager locationManager;
+    private Location lastLocation;
+    private static final String TAG = "LocationService";
+    private static final String TOAST = "service starting";
+    private static String[] permissions = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION};
+
+
+
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+    }
+
+    @SuppressLint("MissingPermission")
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+
+        KrogerLocationCacher.getInstance().makeTokenRequest(this.getBaseContext());
+        final Handler handler = new Handler();
+        handler.postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                //delay for token variable to have new value on return
+                if (intent != null) {
+                    final String action = intent.getAction();
+                    if (action != null) {
+                        if(action == Constants.PERMISSIONS_GRANTED){
+                            locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
+                            locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 1000, 2, new LocationListener() {
+                                @Override
+                                public void onLocationChanged(Location location) {
+                                    //Initially make nearbyLocations request and if location difference to the last location is significant, make a new Kroger API request for new nearby locations
+                                    if (lastLocation == null) {
+                                        lastLocation = location;
+                                        KrogerLocationCacher.getInstance().getNearbyKrogerLocations(location.getLatitude(), location.getLongitude(), getBaseContext()).toString();
+                                        return;
+                                    }
+                                    if (lastLocation.distanceTo(location) > 3000) {
+                                        KrogerLocationCacher.getInstance().getNearbyKrogerLocations(location.getLatitude(), location.getLongitude(), getBaseContext()).toString();
+                                    }
```

Comment:
we should also update `lastLocation = location` here right?
Also, if we do need to do that, we can combine the 2 if statements with an || operator